src/alexandria3k/data_sources/uspto.py

```diff
@@ -18,71 +18,621 @@
+        return self.filename
+
+
+class VTSource:
+    """Virtual table data source.  This gets registered with the apsw
+    Connection through createmodule in order to instantiate the virtual
+    tables."""
+
+    def __init__(self, data_directory, sample):
+        self.data_files = ZipFiles(data_directory, sample)
+        self.table_dict = {t.get_name(): t for t in tables}
+        self.sample = sample
+
+    def Create(self, _db, _module_name, _db_name, table_name):
+        """Create the specified virtual table
+        Return the tuple required by the apsw.Source.Create method:
+        the table's schema and the virtual table class."""
+        table = self.table_dict[table_name]
+        return table.table_schema(), StreamingCachedContainerTable(
+            table,
+            self.table_dict,
+            self.data_files,
+            self.sample,
+        )
+
+    Connect = Create
+
+    def get_container_iterator(self):
+        """Return an iterator over the data files identifiers"""
+        return self.data_files.get_container_iterator()
+
+    def get_container_name(self, fid):
+        """Return the name of the file corresponding to the specified fid"""
+        return self.data_files.get_container_name(fid)
+
+
+class ChunkCursor(ItemsCursor):
+    """A virtual table cursor over patents data.
+    If it is used through data_source partitioned data access
+    within the context of a ZipFiles iterator,
+    it shall return the single element of the ZipFiles iterator.
+    Otherwise it shall iterate over all elements."""
+
+    def __init__(self, table):
+        """Not part of the apsw VTCursor interface.
+        The table argument is a StreamingTable object"""
+        super().__init__(table)
+        # The following get initialized in Filter()
+        self.container_id = None
+        self.zip_index = None
+        self.current_file_path = None
+        self.xml_contents = []
+
+    def Filter(self, index_number, _index_name, constraint_args):
+        """Always called first to initialize an iteration to the first
+        (possibly constrained) row of the table"""
+
+        if index_number == 0:
+            # No index; iterate through all the files
+            self.zip_index = 0
+            self.container_id = -1
+            self.single_file = False
+            self.current_file_path = (
+                self.table.data_source.get_current_zip_path_by_id(
+                    self.zip_index
+                )
+            )
+        elif index_number & CONTAINER_INDEX:
+            # Index; constraint reading through the specified file
+            self.single_file = True
+            self.file_read = False
+            self.container_id = constraint_args[0] - 1
+            self.current_file_path = (
+                self.table.data_source.get_current_zip_path()
+            )
+        self.Next()
+
+    def Next(self):
+        """Advance reading to the next available file. Files are assumed to be
```

Comment:
Please express this in terms of container, clarifying whether this is a zip file or (much better) a parsed XML chunk.